### ISSUE TITLE
[RFC] Interfaces in Nim

### ISSUE THREAD

**Issue by dom96** [nim-lang/RFCs#39]
I've recently had a discussion about this with @Araq, so this is more of a reminder for myself and him, but comments are still welcome.

Araq's ultimate goal for Nim is to have a simple language with a powerful macro system. As such, interfaces in Nim should be implemented using macros.

**Goals:**

* Compatibility with the "fake" interface used in the ``streams`` module

**Potential further research:**

* Look into some of the many OOP/Interface packages to see how they have been implemented. Perhaps with the idea to adopt them into the stdlib in mind.

**Comment by Menduist** [comment]
I posted a $1000 bounty for this issue: https://app.bountysource.com/issues/68171483-rfc-interfaces-in-nim

This is the feature I miss the most for nim, but it doesn't seem to get much traction, so I'm putting some :)

**Comment by daiyam** [comment]
Interfaces can almost be implemented with `concept` but there are several issues:
- `varargs` don't like concepts
- `@[x, y]` don't detect the common concept

```nim
type
  Human = ref object
    name: string
  Dog = ref object

proc makeNoise(human: Human): string =
  "Hello, my name is " & human.name

proc legs(human: Human): int = 2

proc greet(human: Human, other: string): string =
  "Nice to meet you, " & other

proc makeNoise(dog: Dog): string = "Woof! Woof!"

proc legs(dog: Dog): int = 4

proc greet(dog: Dog, other: string): string = "Woof! Woooof... wof!?"

type
  Animal = concept x
    x.makeNoise() is string
    x.legs() is int
    x.greet(string) is string

assert(Human is Animal, "Human is Animal")
assert(Dog is Animal, "Dog is Animal")

proc interact(animal: Animal) =
  echo animal.makeNoise
  echo animal.greet("James Bond")

# proc interactAll(animals: varargs[Animal]) =
#   for animal in animals:
#     animal.interact()

echo Human is Animal  # true
echo Dog is Animal    # true

when isMainModule:
  var
    me = Human(name: "Andrea")
    bau = Dog()
  
  me.interact
  bau.interact
  
  # for animal in @[me, bau]:
  #   echo "Number of legs: ", animal.legs()

  # interactAll(me, bau)

```

**Comment by konsumlamm** [comment]
> Interfaces can almost be implemented with `concept` but there are several issues:
> 
>   * `varargs` don't like concepts
>   * `@[x, y]` don't detect the common concept

These are the same issue: concepts don't support vtables, they're just constraints. For some concept A, you can't have a value "of type A". If you have a `arg: A` parameter to a proc, that's just sugar for a generic: `arg: T` where `T: A` is a generic type parameter.

**Comment by daiyam** [comment]
With
```nim
type
  MAnimal = ref object of RootObj
  Human = ref object of MAnimal
    name: string
  Dog = ref object of MAnimal

proc makeNoise(human: Human): string =
  "Hello, my name is " & human.name

proc makeNoise(dog: Dog): string = "Woof! Woof!"

type
  CAnimal = concept x
    x.makeNoise() is string
```
If we need `type Animal = vtref CAnimal` to transform the concept as the type, then why currently, the following code is working:
```nim
assert(Human is MAnimal, "Human is Animal")
assert(Dog is MAnimal, "Dog is Animal")
assert(Human is CAnimal, "Human is Animal")
assert(Dog is CAnimal, "Dog is Animal")

proc interact(animal: CAnimal) =
  echo animal.makeNoise
```
In those examples, `CAnimal` has the same behavior as a referenced object's type.

So I don't see any need for vtables/`vtref`. It has been at least 5 years since that feature is been discussed, it need to move forward.
Or I did misunderstood/missing something, maybe.

- `proc interactAll(animals: varargs[Animal])` has to be supported since `proc interact(animal: Animal)` is supported.

- > @[x, y] don't detect the common concept

We can't expect the compiler to test all the concepts to find the common concept. We need a way to tell that this type or that type are supporting this or that concept.
A solution would to use the keyword `implements` to do something like:
```nim
implements CAnimal for Human:
  proc makeNoise(human: Human): string =
    "Hello, my name is " & human.name

proc makeNoise(dog: Dog): string = "Woof! Woof!"

implements CAnimal for Dog
```

`implements` would validate the type against the concept and add the link to a table or map so that `@[x, y]` can be supported.
(I would prefer `impl` but it isn't a keyword)

I'm ok to implement those changes.

**Comment by konsumlamm** [comment]
Since this hasn't been mentioned yet, some prior art (in order of relevance for Nim):
* Go has interfaces, which can be used as constraints for generics, but can also be automatically converted to a vtable (see e.g. https://go.dev/tour/methods/9 and the following examples). They're implemented implicitly, just like Nim's concepts.
* Rust has traits, which can also be used as constraints for generics or converted to vtables by making them into a "trait object" (see e.g. https://doc.rust-lang.org/book/ch17-02-trait-objects.html). Traits need an explicit implementation though.
* Java (and other OOP languages, like C#, D, Dart, ...) support interfaces, which work like inheritance (every class is a vtable anyway), so you can also have values of some interface type. These languages require to specify the interfaces a class implements when defining it, however.

**Comment by daiyam** [comment]
> Because you are using inheritance?

@andreaferretti No, because [my previous example](https://github.com/nim-lang/RFCs/issues/39#issuecomment-1305867917) is also working and it isn't using inheritance.

**Comment by daiyam** [comment]
@konsumlamm Let's say we have `type VAnimal = vtype CAnimal`, what will be the difference between `proc interact(animal: CAnimal)` and `proc interact(animal: VAnimal)`? `CAnimal` is implicitly used as a type.

In the following code, `CAnimal` is also implicitly used as a type (result as comment):
```
var
  me = Human(name: "Andrea")
  bau = Dog()

echo me is MAnimal # true
echo bau is MAnimal # true
echo me is CAnimal # true
echo bau is CAnimal # true
```
> You can't have a variable of type CAnimal, because it's not an actual type, just a constraint.

Oh, I understand that.

> A solution to what problem exactly?

To have interfaces in Nim. Interfaces are essentially concepts limited to functions and fields.

Interfaces or traits are used to shared common features between objects. This is why I tried to do with `@[x, y]`. I know it can't work now but it should with interface or something similar (concept).

What I'm confused is that `CAnimal` can be used as a type in some cases (`proc interact(animal: CAnimal)` or `echo me is CAnimal`) but not in some other cases (`proc interactAll(animals: varargs[CAnimal]`).
Also, internally, concepts are called `UserTypeClass`

**Comment by daiyam** [comment]
@konsumlamm My `implements CAnimal for Human:` is inspired from Rust which I find fitting for Nim.

**Comment by Menduist** [comment]
There seem to be a lot of confusion between runtime & compile time here
Basically, if your examples currently works, it means that the compiler is smart enough to find the value at compile time.

This is how it would work under the hood
```nim
block:
  proc greet(a: CAnimal): string = "Animal: " & a.makeNoise()
  echo dog.greet()
  # compiled to
  proc greet[T: CAnimal](a: T): string = "Animal: " & a.makeNoise()
  echo dog.greet[Dog]()

# A VAnimal would look like this instead
block:
  type
    VAnimal = vtype CAnimal
    # Compiled to ---
    VAnimal = object
      makeNoise: proc: string
  converter toVAnimal(canimal: CAnimal): VAnimal =
    result(makeNoise: proc: string = canimal.makeNoise())
  # -------
  proc greet(a: VAnimal): string = "Animal: " & a.makeNoise()

  echo dog.greet()
  # Compiled to (thanks to the converter)
  echo greet(toVAnimal(dog))

**Comment by konsumlamm** [comment]
> @konsumlamm Let's say we have `type VAnimal = vtype CAnimal`, what will be the difference between `proc interact(animal: CAnimal)` and `proc interact(animal: VAnimal)`? `CAnimal` is implicitly used as a type.

The difference would be in how the parameter is represented. A `CAnimal` parameter is actually a generic parameter, so Nim generates a new version of the proc for every type you use it with and you just pass the concrete value to the respective version (the actual type is known at compile-time). A `VAnimal` otoh would be a vtable (a pointer to the table of methods) and a pointer the concrete value (or one pointer for both).

> In the following code, `CAnimal` is also implicitly used as a type (result as comment):
> 
> ```
> var
>   me = Human(name: "Andrea")
>   bau = Dog()
> 
> echo me is MAnimal # true
> echo bau is MAnimal # true
> echo me is CAnimal # true
> echo bau is CAnimal # true
> ```

It's not used as a type, concepts just also work with `is`.

> > You can't have a variable of type CAnimal, because it's not an actual type, just a constraint.
> 
> Oh, I understand that.
> 
> > A solution to what problem exactly?
> 
> To have interfaces in Nim. Interfaces are essentially concepts limited to functions and fields.

But you don't need an explicit `implements` for that.

> Interfaces or traits are used to shared common features between objects. This is why I tried to do with `@[x, y]`. I know it can't work now but it should with interface or something similar (concept).
> 
> What I'm confused is that `CAnimal` can be used as a type in some cases (`proc interact(animal: CAnimal)` or `echo me is CAnimal`) but not in some other cases (`proc interactAll(animals: varargs[CAnimal]`). Also, internally, concepts are called `UserTypeClass`

It's used as a type in none of these cases, the first is just sugar for generics and the second is builtin.

**Comment by daiyam** [comment]
>  the library-based one seems too limited

@Menduist What are you missing in the library [interfaced](https://github.com/andreaferretti/interfaced)?
